src/main/scala/com/evolutiongaming/crypto/Crypto.scala

```diff
@@ -1,149 +1,226 @@
-    new SecretKeySpec(raw, algorithm)
+  private object AES_V3 {
+    /*
+    implementation based on
+    https://proandroiddev.com/security-best-practices-symmetric-encryption-with-aes-in-java-7616beaaade9
+     */
+
+    private val CipherTransformation = "AES/GCM/NoPadding"
+    /*
+    same auth tag length as in
+    https://proandroiddev.com/security-best-practices-symmetric-encryption-with-aes-in-java-7616beaaade9
+     */
+    private val AuthTagLengthBits = 128
+    /*
+    for GCM IV a 12 byte random byte-array is recommend by NIST
+    https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
+     */
+    private val CurrentIVLengthBytes = 12
+    private val MinIVLengthBytes = 12 //does not allow decrypting with IVs smaller than 12 bytes (96 bits)
+
+    def encrypt(value: String, privateKey: String): String = {
+      val skeySpec = aesSKey128bitWithSha256(privateKey.getBytes(UTF_8))
+      val iv = new Array[Byte](CurrentIVLengthBytes)
+      secureRandom.nextBytes(iv)
+      val parameterSpec = new GCMParameterSpec(AuthTagLengthBits, iv)
+      val cipher = Cipher.getInstance(CipherTransformation)
+      cipher.init(Cipher.ENCRYPT_MODE, skeySpec, parameterSpec)
+      val encryptedValue = cipher.doFinal(value.getBytes(UTF_8))
+      encodeEncryptedToString(iv, encryptedValue)
+    }
+
+    private def encodeEncryptedToString(iv: Array[Byte], encryptedValue: Array[Byte]): String = {
+      //1 byte for dynamic IV length encoding
+      val buf = ByteBuffer.allocate(1 + iv.length + encryptedValue.length)
+      require(iv.length >= MinIVLengthBytes, s"IV length shouldn't be smaller than $MinIVLengthBytes bytes")
+      //encode IV length as an unsigned byte starting from MinIVLengthBytes
+      buf.put((iv.length - MinIVLengthBytes).toByte)
+      buf.put(iv)
+      buf.put(encryptedValue)
+      Base64.encodeBase64String(buf.array())
+    }
+
+    def decrypt(value: String, privateKey: String): String = {
+      val payload = Base64.decodeBase64(value)
+      val skeySpec = aesSKey128bitWithSha256(privateKey.getBytes(UTF_8))
+      val cipher = Cipher.getInstance(CipherTransformation)
+      val ivLength = readValidIvLength(payload)
+
+      val ivOffset = 1 //1 byte for encoded IV length
+      val ivEndIdx = ivOffset + ivLength
+      if (payload.length < ivEndIdx) {
+        throw new IllegalArgumentException("invalid data size")
+      }
+      val gcmParamSpec = new GCMParameterSpec(AuthTagLengthBits, payload, ivOffset, ivLength)
+      cipher.init(Cipher.DECRYPT_MODE, skeySpec, gcmParamSpec)
+      val decryptInputLength = payload.size - ivEndIdx
+      val decryptedValue = try {
+        cipher.doFinal(payload, ivEndIdx, decryptInputLength)
+      } catch {
+        case e: AEADBadTagException => throw new DecryptAuthException(e)
+      }
+      new String(decryptedValue, UTF_8)
+    }
+
+    private def readValidIvLength(payload: Array[Byte]): Int = {
```

Comment:
This will be not needed if the length is stored inside encrypted value as is. A bit less magic.

Reply:
Ok to keep it as additional sanity check.